[run.sh] Add retry if HTTP status code != 200

baylanger opened the pull request:
Following the add retry if no IP is associated to the DOMAIN, I found another case that can fail.

For example, when using a Cloudflare tunnel until the tunnel is in ready state, the HTTP status code will be 502. Just earlier I faced this issue, my guess is that others could as well.

This PR adds:
1) a retry until a 2xx code is returned
2) add curl option -L to follow a redirect - this is just in case someone has an odd config and doesn't break anything.

Let me know what you think / if need to adjust anything.

Thank you.

baylanger commented:
Tesla integration has a retry to the proxy , at boot I noticed it says something like "Authentication failed, will retry" but it also have a timeout. When the timeout is reached, it shows the err "Authentication expired". I dislike this behavior :/ - when this condition happens, the end user is forced to reconfigure everything :(

I wonder if the Tesla http proxy accept the requests and return a "retry later status code" ? or should the Tesla integration, when using the proxy, keeps trying instead of giving up and forcing the end user to start all over.

baylanger commented:
I had forgotten to test with a bad request , oops.

Did tests with a valid URL and invalid one to fetch the key, works great.

baylanger commented:
lol found why right at the top :D 

```#!/command/with-contenv bashio
set -e
```

baylanger commented:
Sorry for the spam, latest commit retries doesn't match exactly 200.

llamafilm commented on tesla_http_proxy/rootfs/app/run.sh:
```diff
@@ -89,10 +89,24 @@ done
-  bashio::log.fatal "Fix public key before proceeding."
-  exit 1
-fi
+set +e
```
I don't like changing `set` midway through a script, it makes debugging more confusing.  Instead, can you add `|| true` to the end of curl?

baylanger replied:
I actually thought about this after the last commit. Will change it.

llamafilm commented:
I'm not sure the `-L` is a good idea.  Have you tested to see if the Tesla vehicle follows redirects?

llamafilm commented:
I don't have time to test this right now, but my first thought is, why does your HTTP response code change often enough to justify a retry?  I would expect that if you get something other than 200, that's a problem you need to address manually, so it makes sense to just fail with an error.

baylanger commented:
> I don't have time to test this right now, but my first thought is, why does your HTTP response code change often enough 
> to justify a retry? I would expect that if you get something other than 200, that's a problem you need to address manually, 
> so it makes sense to just fail with an error.
>
I guess there could be a few edge cases, I forgot to mention in the 1st comment, when HAOS is installed as a VM on a NAS and the tunnel is configured in a docker instances, if HAOS comes up before the docker instance tunnel, Cloudflare frontend will return a 502.... if someone uses the HAOS Cloudflare add-on, I'm not sure if the tunnel comes up before the tesla-proxy? If not, that could also make the request for the pubkey fail.

I agree it's something to address manually, same if there's no IP address associated with the DOMAIN, right?

In the PR the http status code is sent to the console (log), it might help someone understand what's wrong - backend issue (5xx) or nginx not properly configured / bad key location (4xx).

baylanger commented:
> I'm not sure the `-L` is a good idea. Have you tested to see if the Tesla vehicle follows redirects?
>
True true, I didn't test if it follows redirects... I don't have the time to put such a setup in place and test it. Makes sense, I just push a commit to remove it.

baylanger commented:
Just fixed the IF 200 or not logic, not sure how I had made such erro.. maybe my AHDH.

Below the output from 3 tests.

If you prefer, we could change the logic to loop X times and then give up. I tried to find the logic in Tesla integration that reports "Authentication expired" and I could not. If the proxy never comes up and the T integration is already configured, I wish it would not fail but wait for the proxy to come up. By failing, the end user needs to reconfigure everything... sadly I can't find where that logic is to propose perhaps a change.

1) Test with return 200
```
[11:54:30] INFO: Testing public key...
HTTP/2 200
date: Thu, 23 May 2024 15:54:30 GMT
content-type: text/plain
content-length: 178
accept-ranges: bytes
etag: "664c346a-b2"
last-modified: Tue, 21 May 2024 05:43:06 GMT
strict-transport-security: max-age=2592000; includeSubDomains; preload
cf-cache-status: DYNAMIC
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=242gHErBlGuSP8qVFImrT1FRkmSCL%2BCXYrujGEuNm913%2FAPOTR2ihwibP%2FuglMDHkgcZG8rgubxUZx%2FyCZZ6iGhAhmukTdW%2BgIlP0Ji1CVOJ85tO9%2BudXkWRUpIHUP7gDcxXoAF%2FwNkA80OV"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-content-type-options: nosniff
server: cloudflare
cf-ray: 888639958bf43905-YYZ
alt-svc: h3=":443"; ma=86400

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEkD6ShEF+P7/d02A8HUevVlwtKwqc
UmQXrSTuMdLXWcV17I6mTqVMgJKSP7L7QeJofMjKQ0Tuu1AS/hPNn+0s5A==
-----END PUBLIC KEY-----
[11:54:30] INFO: The public key is accessible.
```

2) public key not found (404)
```
[12:01:50] INFO: Testing public key...
HTTP/2 404
date: Thu, 23 May 2024 16:01:51 GMT
content-type: text/html
strict-transport-security: max-age=2592000; includeSubDomains; preload
cf-cache-status: DYNAMIC
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=pnjECrdCRAvPt8AUKHvsQ4JOPZvaps1k8pHVKtL2tSxsetn2chn4qc%2BLzWKzxLIR3%2Fd1GPHwd6pW4FV%2BKhUv5Hg%2BwZOdkOdWbsjllixn50LdE6suilD5qvoWDD8yzbti%2FRHRwRGQDjfWIg1R"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-content-type-options: nosniff
server: cloudflare
cf-ray: 8886445589cf37cf-YYZ
alt-svc: h3=":443"; ma=86400

[12:01:51] FATAL: HTTP status code 404; Use a search engine to learn about the status code.
[12:01:51] FATAL: If the request keeps failing, adjust your configuration for the request not to fail.
[12:01:51] FATAL: Sleeping 2 minutes before retry.
```

3) Cloudflare tunnel down
```
[12:03:15] INFO: Testing public key...
HTTP/2 530
date: Thu, 23 May 2024 16:03:16 GMT
content-type: text/plain; charset=UTF-8
content-length: 16
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=39q3K58Q%2BeSfqEuHjkPBz922l36xcLrGDWhzaTTuPOdeNNmJulXBm2Kyn6ZEnqgmy2wVdd0MpEOuoFVlBw%2FDIOmn6FpVoHGV3QH9NwTFdGp%2F7YtWKvNTp5mQ%2Bwz1dzwRxSt6xALW8p7LMpTD"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
strict-transport-security: max-age=2592000; includeSubDomains; preload
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
referrer-policy: same-origin
cache-control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
expires: Thu, 01 Jan 1970 00:00:01 GMT
server: cloudflare
cf-ray: 888646691c797117-YYZ
alt-svc: h3=":443"; ma=86400

[12:03:16] FATAL: HTTP status code 530; Use a search engine to learn about the status code.
[12:03:16] FATAL: If the request keeps failing, adjust your configuration for the request not to fail.
[12:03:16] FATAL: Sleeping 2 minutes before retry.
```

llamafilm commented:
Good points — I have not tested rebooting the Home Assistant box to see which comes up first, and if the order varies.  I think it makes sense to loop forever as you have done.

Sorry for being picky, but I think this is needlessly verbose.  You could consolidate those 3 error lines into 1 like this:
```
[12:01:51] ERROR: HTTP status code 404; Sleeping 2 minutes before retry.
```

And change fatal to error.

baylanger commented:
No you're not picky at all, you don't know me enough ;)

I don't mind compressing, I went verbose mainly for "newbies" - Done in last commit.

Changed from fatal to error , I wasn't too sure what bashio supported for errors... - Done
I also changed fatal to error from my previous PR that you merged ~ last week.

llamafilm commented:
Thank you!
